presto-thrift-tpch-node/src/main/java/com/facebook/presto/thrift/node/ThriftServerTpch.java

```diff
@@ -0,0 +1,462 @@
+        List<Type> keyTypes = types(indexInfo.getTableName(), table.getKeyColumns());
+        Page keysPage = convertToPage(keys, table.getKeyColumns(), keyTypes);
+        RecordSet keyRecordSet = new PageRecordSet(keyTypes, keysPage);
+        RecordSet allColumnsOutputRecordSet = table.lookupKeys(keyRecordSet);
+        List<Integer> outputRemap = computeRemap(table.getOutputColumns(), indexInfo.getOutputColumnNames());
+        RecordSet outputRecordSet = new MappedRecordSet(allColumnsOutputRecordSet, outputRemap);
+        return new ThriftSplitsOrRows(
+                null,
+                cursorToRowsBatch(
+                        outputRecordSet.cursor(),
+                        indexInfo.getOutputColumnNames(),
+                        estimateRecords(rowsMaxBytes, outputRecordSet.getColumnTypes()),
+                        continuationToken));
+    }
+
+    private ThriftRowsBatch getRowsInternal(byte[] splitId, long maxBytes, @Nullable byte[] continuationToken)
+    {
+        SplitInfo splitInfo = deserialize(splitId, SplitInfo.class);
+        List<String> columnNames = splitInfo.getColumnNames();
+        RecordCursor cursor = createCursor(splitInfo, splitInfo.getColumnNames());
+        return cursorToRowsBatch(
+                cursor,
+                columnNames,
+                estimateRecords(maxBytes, types(splitInfo.getTableName(), splitInfo.getColumnNames())),
+                continuationToken);
+    }
+
+    private static int estimateRecords(long maxBytes, List<Type> types)
+    {
+        if (types.isEmpty()) {
+            // no data will be returned, only row count
+            return Integer.MAX_VALUE;
+        }
+        int bytesPerRow = 0;
+        for (Type type : types) {
+            if (type instanceof FixedWidthType) {
+                bytesPerRow += ((FixedWidthType) type).getFixedSize();
+            }
+            else if (type instanceof VarcharType) {
+                VarcharType varchar = (VarcharType) type;
+                if (varchar.isUnbounded()) {
+                    // random estimate
+                    bytesPerRow += 32;
+                }
+                else {
+                    bytesPerRow += varchar.getLengthSafe();
+                }
+            }
+            else {
+                throw new IllegalArgumentException("Cannot compute estimates for an unknown type: " + type);
+            }
+        }
+        return toIntExact(maxBytes / bytesPerRow);
+    }
+
+    private static ThriftRowsBatch cursorToRowsBatch(
+            RecordCursor cursor,
+            List<String> columnNames,
+            int maxRowCount,
+            @Nullable byte[] continuationToken)
+    {
+        long skip = continuationToken != null ? Longs.fromByteArray(continuationToken) : 0;
+        // very inefficient implementation as it needs to re-generate all previous results to get the next batch
+        skipRows(cursor, skip);
```

Comment:
`TpchRecordSet` support "parts" which is not exactly a pagination. So, you cannot resume from a certain position within a part. Typically, parts are mapped to splits: one split - one part.
Technically, it should be possible to pre-split a table to have several parts per split, so that amount of data that needs to be generated again is minimal or even avoid it altogether, but it requires some assumption about requested data size and is slightly harder to implement.
The current implementation should be good enough for tests, so I'd like to not be blocked on it now. I'll think about implementing it without skipping records.